parglare/parser.py

```diff
@@ -474,36 +490,32 @@ def _next_token(self, context):
 
         in_len = len(input_str)
 
-        # Find the next token in the input
-        if position == in_len and EMPTY not in actions \
-           and STOP not in actions:
-            # Execute EOF action at end of input only if EMPTY and
-            # STOP terminals are not in actions as this might call
-            # for reduction.
-            ntok = EOF_token
-        else:
-            tokens = []
-            if position < in_len:
-                if self.custom_token_recognition:
+        tokens = []
 
-                    def get_tokens():
-                        return self._token_recognition(context)
+        # add special tokens (EMPTY, STOP and EOF) if the are applicable
+        if EMPTY in actions:
+            tokens.append(EMPTY_token)
+        if STOP in actions:
+            tokens.append(STOP_token)
+        if position == in_len:
+            tokens.append(EOF_token)
+
+        # Get tokens by trying recognizers
+        if self.custom_token_recognition:
+            def get_tokens():
+                return self._token_recognition(context)
+
+            tokens.extend(self.custom_token_recognition(
+                context, get_tokens,
+            ))
```

Comment:
Here `custom_token_recognition` is required to return iterable. I would check if there is anything returned before calling `tokens.extend` as it would make it more consistent with recognizers and more backward compatible, not to mention boilerplate `return []` in the `custom_token_recognition`.

Reply:
Requirement of returning iterable from `custom_token_recognition` is a good thing in my opinion. This way we are decreasing complexity on the library side and enforcing clean return type from a function - return value must be an iterable without any exceptions. Meanwhile we are not increasing complexity on users side. We are increasing their line count a bit, but not complexity. `return []` is a boilerplate, but not particularly bad, because it is as much complex as bare `return None`, and the only reason `return None` is considered more compact is due to python's syntactic sugar.

Anyway, if it's blocking you from merging this I will change the code according to your recommendations. Let me know.

Reply:
Recognizers are quite different sort of function so making`custom_token_recognition` compatible with them is not a thing I consider required. [Recognizer is a function that matches on input string and returns a match or no match and `custom_token_recognition` matches on input and returns set (iterable) of possible matches.]

As for backwards compatibility - if it must be preserved then I'll adjust. However if it doesn't then I suggest changing the behaviour right now to make API more strict. (By strict I mean it verifies user's input more.)

Reply:
I understand your reasoning but I would like to make it work like before for pragmatic reasons. I would like to make it easier for the user. It is idiomatic in Python to not handle cases where `None` should be returned and as you correctly observed it is a Python syntactic sugar for `return None` but Python programmers are used to this way of work. So, you just have to handle cases you care about in the function and leave other branches "empty". Insisting on `return []` on unsuccessful matches is not idiomatic Python and puts additional burden on lib users. And in tight loops it creates more overhead as creating a list is more expensive than returning None. I would say that your background is in functional languages where this line of reasoning is the norm (and BTW I do agree from the theoretical point of view :) ).

Recognizers are not much different as they receive a sequence and should return a matched subsequence from the current position. That might be a string but also can be a sequence of any other objects. So, we could reason here the same way, since we should return a matched subsequence it should be an empty sequence/list if nothing matches but we return `None` for pragmatic reasons.